### PR TITLE
XOR gate example

### PR DESCRIPTION
**Pull request by neoneye**
I have tried training a network to behave as an xor gate, but I cannot get any meaningful output.

Here is the output that I'm seeing in the console.

	before training: [0.23279, -0.223946, -0.366412, -0.566801, -0.022422, 0.311922, 0.352381, 0.373161, -0.2297]
	after training: [0.257579, -0.249485, -0.405368, -0.629288, -0.0246187, 0.345451, 0.380216, 0.406903, -0.257867]
	values: 
	[0.627165] [0.612517] [0.621583] [0.606965]
	Program ended with exit code: 0

Your code looks great.

### CODE REVIEW

**Comment by collinhundley** [comment]
Ok, the problem is that you haven't given the trainer any validation data, which was causing it to exit after the very first training epoch. This was also a flaw in my own design - I'm now adding a check that will generate an error if an empty validation set is provided.

Here's a modified version of your test that works well:
```
try! network.train(inputs: inputs, answers: answers,
                    testInputs: inputs, testAnswers: answers,
                    errorThreshold: 0.0001)
```
Notice that I'm just passing in `inputs` and `answers` as the validation data as well - since there are only four possible inputs anyway, it makes sense to do that here.  In face, perhaps a good solution is to make the validation sets optional, and assume the user wants to validate using the training data if none are provided.

Also notice that I set the error threshold to `0.0001`. The previous value of `0.1` was allowing the trainer to exit before it converged on a good solution.  Here's my output:

```
before training: [0.560786, 0.542522, 0.267538, -0.106166, -0.530252, -0.025325, -0.340144, 0.477577, 0.350304]
after training: [-2.7255, -0.0987173, 5.35911, -2.38694, -0.27567, 4.87529, -6.00387, 6.76224, 5.65316]
values: 
[0.00598653] [0.99603] [0.00518748] [0.995309]
```

I hope to come up with a formula to automatically calculate a reasonable error threshold based on the input data provided.  Right now it's just trial-and-error.

**Comment by neoneye** [comment]
Optional validation sets. That's sounds like a good idea, although I'm not a machine learning expert.

Now it works! Thank you Collin for great help :+1:

**Comment by collinhundley** [comment]
Thanks for contributing! I've merged your example into the master branch, with a couple notes added.
